src/numbers/xdrint.js

```diff
@@ -0,0 +1,246 @@
+import { Uint256 } from './uint256';
+import { Int128 } from './int128';
+import { Int256 } from './int256';
+
+import xdr from '../xdr';
+
+export class XdrInt {
+  int; // child class of a jsXdr.LargeInt
+  type; // string, one of i64, u64, i128, u128, i256, or u256
+
+  /**
+   * Transforms an opaque {@link xdr.ScVal} into a native BigInt, if possible.
+   *
+   * You can then give this back to create an {@link ScInt} instance, but the
+   * rationale here is that the native type is more likely to be immediately
+   * useful.
+   *
+   * @param {xdr.ScVal} scv - the raw XDR value to parse into an integer
+   * @returns {bigint} the integer value, regardless of size (even 32-bit)
+   * @throws {TypeError} if the input value doesn't represent an integer
+   */
+  static fromScVal(scv) {
+    switch (scv.switch().name) {
+      case 'scvU32':
+      case 'scvI32':
+        return BigInt(scv.value());
+
+      case 'scvU64':
+      case 'scvI64':
+        return scv.value().toBigInt();
+
+      case 'scvU128':
+        return new Uint128(scv.value().lo(), scv.value().hi()).toBigInt();
+
+      case 'scvI128':
+        return new Int128(scv.value().lo(), scv.value().hi()).toBigInt();
+
+      case 'scvU256':
+        return new Uint256(
+          scv.value().loLo(),
+          scv.value().loHi(),
+          scv.value().hiLo(),
+          scv.value().hiHi()
+        ).toBigInt();
+
+      case 'scvI256':
+        return new Int256(
+          scv.value().loLo(),
+          scv.value().loHi(),
+          scv.value().hiLo(),
+          scv.value().hiHi()
+        ).toBigInt();
+
+      default:
+        throw TypeError(`expected integer type, got ${scv.switch()}`);
+    }
+  }
+
+  constructor(type, values) {
+    if (!(values instanceof Array)) {
+      values = [values];
+    }
+
+    // normalize values to one type
+    values = values.map((i) => {
+      // micro-optimization to no-op on the likeliest input value:
+      if (typeof i === 'bigint') {
+        return i;
+      }
+      if (i instanceof XdrInt) {
+        return i.toBigInt();
+      }
+      return BigInt(i);
+    });
+
+    switch (type) {
+      case 'i64':
+        this.int = new Hyper(values);
+        break;
+      case 'i128':
+        this.int = new Int128(values);
+        break;
+      case 'i256':
+        this.int = new Int256(values);
+        break;
+      case 'u64':
+        this.int = new UnsignedHyper(values);
+        break;
+      case 'u128':
+        this.int = new Uint128(values);
+        break;
+      case 'u256':
+        this.int = new Uint256(values);
+        break;
+      default:
+        throw TypeError(`invalid type: ${type}`);
+    }
+
+    this.type = type;
+  }
+
+  /**
+   * @returns {number}
+   * @throws {RangeError} if the value can't fit into a Number
+   */
+  toNumber() {
+    const bi = this.int.toBigInt();
+    if (bi > Number.MAX_SAFE_INTEGER || bi < Number.MIN_SAFE_INTEGER) {
+      throw RangeError(`value ${bi} too large for Number`);
```

Comment:
Maybe two errors, one for too large, one for too small?

Reply:
Fixed it by including the range itself in the error! ee42e30